RHIDP-5494: Doc air-gapped installation using Operator (#935)
redhat-developer/red-hat-developers-documentation-rhdh: modules/installation/proc-install-rhdh-airgapped-environment-ocp-operator.adoc

* If you are using an {ocp-short} cluster, you have the following prerequisites:
** (Optional) You have installed the `oc-mirror` {ocp-short} CLI plugin if you want to use it to mirror images.
* If you are using a supported Kubernetes cluster, you have the following prerequisites:
** You have installed the Operator Lifecycle Manager (OLM) on the disconnected cluster.
** You have a mirror registry that is reachable from the disconnected cluster.

.Procedure
. In your terminal, navigate to the directory where you want to save the mirroring script.
. Download the mirroring script by running the following command:
+
[source,terminal,subs="attributes+"]
----
curl -sSLO https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/release-{product-version}/.rhdh/scripts/prepare-restricted-environment.sh
----
+
. Run the mirroring script by running the `bash` command with the appropriate set of options:
+
[source,terminal,subs="+quotes,+attributes"]
----
bash prepare-restricted-environment.sh \ (1)

>   [--to-registry _<my.registry.example.com>_]] \ (2)

Reviewer: ```suggestion
  --filter-versions "{product-version}" \
  [--to-registry _<my.registry.example.com>_] \ (2)
```

I realized that the default version installed is 1.4 in the script. The docs need to reflect the right version that the user will use (even if 1.5 is currently not GA yet).

Author: You're right. We'll actually use _<product_version>_ in the code to indicate replaceable text where the user can insert whichever product version they want to use and we'll use the attribute in the variable definition/example.

Reviewer: @linfraze Has this been added? I might have missed it.

Author: @rm3l had to switch tasks for a while but this update has been added now

  [--use-oc-mirror true]
----
<1> Specifies the URL for the target mirror registry where you want to mirror the images.
<2> (Optional) Uses the `oc-mirror` {ocp-short} CLI plugin to mirror images.
+
[NOTE]
====
The script can take several minutes to complete as it copies multiple images to the mirror registry.
====

.Verification
* If you are using {ocp-brand-name}, the {product} Operator is in the *Installed Operators* list in the web console.
* If you are using a supported Kubernetes platform, you can check the list of pods running in the `rhdh-operator` namespace by running the following command in your terminal:
+
[source,terminal,subs="+quotes,+attributes"]
----
kubectl -n rhdh-operator get pods

> ----

Reviewer: The helper script installs the Operator only (not RHDH). It is then up to the user to install RHDH by creating a CR. Since the goal of this page is ultimately "Installing Red Hat Developer Hub in an air-gapped environment", do you want to add further instructions on how to install RHDH once the Operator is up and running (or link to other procedures)?

Author: @rm3l This is a Q I had - is the rest of the disconnected procedure the same as installing RHDH (on whichever supported platform the customer is using) on a connected network? If yes, then we can link to those existing procedures in a **Next steps** section. If the disconnected process requires additional/different steps, then we will need to include those here.

Author: > if the mirror registry requires authentication.

so there are only certain conditions under which authentication is required? what are those conditions?

Reviewer: @linfraze This depends on how the environment is set up. But I assume most mirror registries would require some authentication (at least for registries reachable from the internet, like in partially disconnected environments).

Author: Created a Jira issue to add a conditional prereq to deployment docs - https://issues.redhat.com/browse/RHIDP-6367

* You have an active `oc registry` session to the `registry.redhat.io` {company-name} Ecosystem Catalog. For more information, see link:https://access.redhat.com/RegistryAuthentication[{company-name} Container Registry Authentication].
* You have installed the `opm` CLI tool. For more information, see link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/cli_tools/opm-cli#olm-about-opm_cli-opm-install[Installing the opm CLI].
* You have installed Podman 5.3 or later. For more information, see link:https://podman.io/docs/installation[Podman Installation Instructions].

.Procedure
. Download the mirroring script to disk by running the following command:
+
[source,terminal,subs="attributes+"]
----
curl -sSLO https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/release-{product-version}/.rhdh/scripts/prepare-restricted-environment.sh
----
+
. Run the mirroring script by running the `bash` command with the appropriate set of options:
+
[source,terminal,subs="+quotes,+attributes"]
----
bash prepare-restricted-environment.sh \

>     --to-dir _<my_pulled_image_location>_

Reviewer: ```suggestion
    --filter-versions "{product-version}" \
    --to-dir _<my_pulled_image_location>_ \
    [--use-oc-mirror true]
```

----
curl -sSLO https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/release-{product-version}/.rhdh/scripts/prepare-restricted-environment.sh
----
+
. Run the mirroring script by running the `bash` command with the appropriate set of options:
+
[source,terminal,subs="+quotes,+attributes"]
----
bash prepare-restricted-environment.sh \
 --filter-versions "{product-version}" \
  [--to-registry _<my.registry.example.com>_] \ <1>
  [--use-oc-mirror true] <2>
----
<1> Specifies the URL for the target mirror registry where you want to mirror the images.
<2> (Optional) Uses the `oc-mirror` {ocp-short} CLI plugin to mirror images.
+
[NOTE]
====
The script can take several minutes to complete as it copies multiple images to the mirror registry.
====

.Verification
* If you are using {ocp-brand-name}, the {product} Operator is in the *Installed Operators* list in the web console.
* If you are using a supported Kubernetes platform, you can check the list of pods running in the `rhdh-operator` namespace by running the following command in your terminal:
+
[source,terminal,subs="+quotes,+attributes"]
----
kubectl -n rhdh-operator get pods
----

> include::../../artifacts/snip-rhdh-install-operator-next-steps.adoc[]

Author: @docs - is there an issue with using the same snippet more than once within a single file? it's not showing up in both places and the ccutil build is failing.

Author: removing one of the snippets fixed the build error

* Operands images ({product-very-short}, PostgreSQL)

To make these resources available, replace them with their equivalent resources in a mirror registry accessible to your cluster.

You can use a helper script that mirrors the necessary images and provides the necessary configuration to ensure those images will be used when installing the {product} Operator and creating {product-short} instances. This script requires a target mirror registry. You likely have a target mirror registry ready to use if your cluster is already operating on a disconnected network. If you do not already have a target registry, and if you have an {ocp-short} cluster, you might want to expose and leverage the internal cluster registry.

If you are connected to a {ocp-short} cluster, the helper script will detect it and will automatically expose the cluster registry. However, if you are connected to a Kubernetes cluster, you can manually specify the target registry that you want to mirror the images to.

.Prerequisites
* You have installed Podman 5.3 or later. For more information, see link:https://podman.io/docs/installation[Podman Installation Instructions].
* You have installed Skopeo 1.17 or later.

> * You have installed yq 4.44 or later.

Author: @reestr running the bash script (in place of performing the manual steps required to install the Operator in an air-gapped environment) requires the user to install several tools that (afaik) are not owned or maintained by Red Hat, including `skopeo`, `yq`, `umoci`. Providing any information about these tools or how to install them requires us to link to third-party websites (GitHub repos) that are also not owned or maintained by Red Hat.

What are your thoughts on using bash scripts to perform RHDH installation procedures? Does this approach cross the boundaries of RH support?

** link:https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.4/html/installing_red_hat_developer_hub_on_amazon_elastic_kubernetes_service/proc-rhdh-deploy-eks-operator_title-install-rhdh-eks[Installing {product-short} on {eks-short} with the Operator]
** link:https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.4/html/installing_red_hat_developer_hub_on_microsoft_azure_kubernetes_service/proc-rhdh-deploy-aks-operator_title-install-rhdh-aks[Installing {product-short} on {aks-short} with the Operator]
** link:https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.4/html/installing_red_hat_developer_hub_on_openshift_dedicated_on_google_cloud_platform/proc-install-rhdh-osd-gcp-operator_title-install-rhdh-osd-gcp[Installing {product-short} on {gcp-short} with the Operator]
** link:https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.4/html/installing_red_hat_developer_hub_on_google_kubernetes_engine/proc-rhdh-deploy-gke-operator.adoc_title-install-rhdh-gke#proc-deploy-rhdh-instance-gke.adoc_title-install-rhdh-gke[Deploying {product-short} on {gke-short} with the Operator]

> == Installing {product} in a fully disconnected environment with the Operator

Author: +1 to more "why" info in general, and we can definitely provide some more conceptual information here, but using a fully disconnected environment is often a requirement rather than a choice. Some customers, like government agencies, must use a fully disconnected environment due to their high-security needs and are likely already knowledgeable about their technical environment and why it is needed.

> == Installing {product} in a fully disconnected environment with the Operator

Reviewer 2: We’ve previously received feedback that our documentation often assumes users are already knowledgeable, and we simply present the information. Since security is a key reason users might choose a disconnected environment, how about we include something like:
"_In environments without internet access — whether for security, compliance, or operational reasons — a fully disconnected installation ensures that {product} can run reliably without external dependencies._"

This could help provide context for less experienced users, giving them a clearer understanding of why they might need this setup or why it’s beneficial.

> == Installing {product} in a fully disconnected environment with the Operator

Author: Sure, I'll paste it in

[id="proc-install-rhdh-airgapped-environment-ocp-operator_{context}"]
= Installing {product} in an air-gapped environment with the Operator

You can install {product} in a fully disconnected or partially disconnected environment using the {product} Operator. For a list of supported platforms, see the link:https://access.redhat.com/support/policy/updates/developerhub[{product} Life Cycle page].

> == Installing {product} on {ocp-short} in a partially disconnected environment with the Operator

Author: Good catch - The OCP reference is not supposed to be there and was removed from the main title.

* Operator images (bundle, operator, catalog)
* Operands images ({product-very-short}, PostgreSQL)

To make these resources available, replace them with their equivalent resources in a mirror registry accessible to your cluster.

You can use a helper script that mirrors the necessary images and provides the necessary configuration to ensure those images will be used when installing the {product} Operator and creating {product-short} instances. This script requires a target mirror registry. You likely have a target mirror registry ready to use if your cluster is already operating on a disconnected network. If you do not already have a target registry, and if you have an {ocp-short} cluster, you might want to expose and leverage the internal cluster registry.

If you are connected to a {ocp-short} cluster, the helper script will detect it and will automatically expose the cluster registry. However, if you are connected to a Kubernetes cluster, you can manually specify the target registry that you want to mirror the images to.

.Prerequisites

> * You have an active `oc registry` session to the `registry.redhat.io` {company-name} Ecosystem Catalog. For more information, see link:https://access.redhat.com/RegistryAuthentication[{company-name} Container Registry Authentication].

Author: I think the second `podman` should be `skopeo` so will apply that change

   --prod_operator_index "registry.redhat.io/redhat/redhat-operator-index:v{ocp-version}" \
   --prod_operator_package_name "rhdh" \
   --prod_operator_bundle_name "rhdh-operator" \
   --prod_operator_version "v{product-bundle-version}" \
   --use_existing_mirror_registry "_<my_registry>_"
----
+
[NOTE]
====
The script can take several minutes to complete as it copies multiple images to the mirror registry.
====

.Verification
* If you are using {ocp-brand-name}, the {product} Operator is in the *Installed Operators* list in the web console.
* If you are using a supported Kubernetes platform, you can check the list of pods running in the `rhdh-operator` namespace by running the following command in your terminal:
+
[source,terminal,subs="+quotes,+attributes"]
----
kubectl -n rhdh-operator get pods
----

== Installing {product} on {ocp-short} in a fully disconnected environment with the Operator

If your network has access to the registry through a bastion host or physical disk, you can use the Operator to install {product} by mirroring specified resources and transferring them to your air-gapped environment without any connection to the internet.

> .Prerequisites

Reviewer: There are additional prerequisites (common to both partially and fully disconnected environments) mentioned in https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/main/.rhdh/docs/airgap.adoc:

> * GNU sed. See [GNU sed](https://www.gnu.org/software/sed/).
> 
> * (Optional) oc-mirror (the OpenShift Client mirror plugin) if you want to mirror using oc-mirror on OpenShift. See [Installing the oc-mirror OpenShift CLI plugin](https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html-single/disconnected_environments/index#installation-oc-mirror-installing-plugin_about-installing-oc-mirror-v2).
> 
> * Skopeo 1.17+. See [Installing Skopeo](https://github.com/containers/skopeo/blob/main/install.md).
> 
> * umoci. See [Install](https://github.com/opencontainers/umoci#install).
> 
> * yq v4.44+. See [Install](https://github.com/mikefarah/yq#install).
> 
> * (Required on non-OpenShift clusters) The cluster needs to have the Operator Lifecycle Manager installed and already working in disconnected environments. See [Installing OLM](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/install/install.md#installing-olm).
> 
> * (Required on non-OpenShift clusters) A mirror registry reachable from the cluster.

* You have set up your disconnected environment.

> ** You have mirrored all of the required images to disk.

Author: Ah, good point. I pulled this from the upstream doc, btw.

** You have manually transferred the mirror folder to the network of the disconnected mirror registry.

> ** You have mirrored the images from disk to the target mirror registry in your disconnected environment.

Reviewer: Same point: this will be done by the script.

> ** You have installed the Operator in your disconnected environment.

Reviewer: Same point: this is what the script ultimately does when it is done mirroring from disk to the target mirror registry.

* You have set up your workstation.

> ** You have an active `oc registry` session to the `registry.redhat.io` {company-name} Ecosystem Catalog. For more information, see link:https://access.redhat.com/RegistryAuthentication[{company-name} Container Registry Authentication].

Reviewer: Same point as in one of previous comments.. User can use `oc registry`, `podman login` or `skopeo login`.

** You have installed the `opm` CLI tool. For more information, see link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/cli_tools/opm-cli#olm-about-opm_cli-opm-install[Installing the opm CLI].
** You have installed Podman 5.3 or later. For more information, see link:https://podman.io/docs/installation[Podman Installation Instructions].

.Procedure
. Download the mirroring script to disk by running the following command:
+
[source,terminal,subs="attributes+"]
----
curl -sSLO https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/release-{product-version}/.rhdh/scripts/prepare-restricted-environment.sh
----
+
. Run the mirroring script by running the `bash` command with the appropriate set of options:
+
[source,terminal,subs="+quotes,+attributes"]

> ----

Reviewer: The command you mentioned (using `--ci-index true`) is for CI testing purposes, and should not be part of the product docs. 

Instead, this is what the end-user needs to run to mirror the images to disk:

```
bash prepare-restricted-environment.sh \
    --to-dir _<my_pulled_image_location>_
```

bash prepare-restricted-environment.sh \
    --index-image "quay.io/rhdh/iib:next-v4.18-x86_64" \
    --ci-index true \
    --to-dir _<my_pulled_image_location>_ \
    [--filter-versions '*'] \
    [--use-oc-mirror true]
----
+
where

> _<my_pulled_image_location>_ :: Specifies the directory where you want to pull all of the necessary images with the `--to-dir` option, for example, my.registry.example.com/namespace

Reviewer: ```suggestion
_<my_pulled_image_location>_ :: Specifies the absolute path to a directory where you want to pull all of the necessary images with the `--to-dir` option, for example, /home/user/rhdh-operator-mirror-dir
```

`my.registry.example.com/namespace` sounds more like a registry path, not a directory. And I think we should mention that the path needs to be absolute at the moment.

+
[NOTE]
====
The script can take several minutes to complete as it copies multiple images to the mirror registry.
====
+
. Transfer the directory specified by the `--to-dir` option to your disconnected environment.

> . From a machine in your disconnected environment that has access to both the cluster and the target mirror registry, download the mirroring script from disk by running the following command:

Reviewer: It is not clear from https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/main/.rhdh/docs/airgap.adoc, but there is no need to redownload the script again. In the mirror directory transferred by the user, the script will be there, named as `install.sh`.

+
[source,terminal,subs="attributes+"]
----
curl -sSLO https://raw.githubusercontent.com/redhat-developer/rhdh-operator/refs/heads/release-{product-version}/.rhdh/scripts/prepare-restricted-environment.sh
----
+
. Run the mirroring script by running the `bash` command with the appropriate set of options:
+
[source,terminal,subs="+quotes,+attributes"]

> ----

Reviewer: Based on my previous comment, the command could be:

```
bash _<my_pulled_image_location>_/install.sh  \
    --from-dir _<my_pulled_image_location>_ \
    [--to-registry _<my.registry.example.com>_] \
    [--use-oc-mirror true]
```

bash prepare-restricted-environment.sh \
    --from-dir _<my_pulled_image_location>_ \
    [--to-registry _<my.registry.example.com>_/_<namespace>_] \
    [--use-oc-mirror true]
----
+
where

_<my_pulled_image_location>_ :: Specifies the directory where you want to pull all of the necessary images with the `--to-dir` option.

_<my.registry.example.com>_ :: Specifies the URL for the target mirror registry where you want to mirror the images.

> _<namespace>_ :: Specifies the target namespace where you want to mirror the images.

Reviewer: ```suggestion
```

I think we can remove this part, as it is confusing with a Kubernetes namespace. It is actually supposed to be part of the overall registry URL, but it depends on how each mirror registry handles that. It might not be needed on some registries.
